syntax = "proto3";

import "google/protobuf/timestamp.proto";

package protocol;

option go_package = "./proto;proto";

service Protocol {
  rpc SendTransaction (SendTransactionRequest) returns (TransactionResponse);
  rpc GetTransaction (GetTransactionRequest) returns (TransactionResponse);
  rpc GetBlock (GetBlockRequest) returns (GetBlockResponse);
  rpc Ping (PingRequest) returns (PingResponse);
}

message SignedTransaction {
  string signature = 1;
  string request_id = 2;
  oneof transaction {
    TrackPlays plays = 1000;
    ValidatorRegistration validator_registration = 1001;
    SlaRollup sla_rollup = 1002;
    ManageEntityLegacy manage_entity = 1003;
  }
}

message SendTransactionRequest {
  SignedTransaction transaction = 1;
}

message GetTransactionRequest {
  string txhash = 1;
}

message TransactionResponse {
  string txhash = 1;
  SignedTransaction transaction = 2;
}

message GetBlockRequest {
  int64 height = 1;
}

message GetBlockResponse {
  string hash = 1;
  int64 height = 2;
  repeated SignedTransaction transactions = 3;
}

message TrackPlays {
  repeated TrackPlay plays = 1;
}

message ValidatorRegistration {
  string endpoint = 1;
  string comet_address = 2;
  string eth_block = 3;
  string node_type = 4;
  string sp_id = 5;
}

message TrackPlay {
  string user_id = 1;
  string track_id = 2;
  google.protobuf.Timestamp timestamp = 3;
  string signature = 4;
}

message PingRequest {
}

message PingResponse {
  string message = 1;
}

message SlaRollup {
  google.protobuf.Timestamp timestamp = 1;
  int64 block_start = 2;
  int64 block_end = 3;
  repeated SlaNodeReport reports = 4;
}

message SlaNodeReport {
  string address = 1;
  int32 num_blocks_proposed = 2;
}

message ManageEntityLegacy {
  int64 user_id = 1;
  string entity_type = 2;
  int64 entity_id = 3;
  string action = 4;
  string metadata = 5;
  string signature = 6;
}
